packages/asset-swapper/src/network/utils.ts

```diff
@@ -0,0 +1,211 @@
+import { BigNumber, hexUtils } from '@0x/utils';
+import { FastABI } from 'fast-abi';
+import { ContractVersionData , ContractAbi, MethodAbi, SupportedProvider } from 'ethereum-types';
+
+import { artifacts } from '../artifacts';
+
+import { Chain, ChainEthCallOpts } from './chain';
+import { DUMMY_PROVIDER } from './constants';
+import { Address, Bytes, DexSample, MultiHopCallInfo } from './types';
+
+const ADDRESS_SIZE = 20;
+
+interface ArtifactsMap {
+    [artifactName: string]: ContractVersionData;
+}
+
+export interface ContractWrapperType<T> {
+    new (
+        address: Address,
+        provider: SupportedProvider,
+        _txDefaults: {},
+        _logDeps: {},
+        deployedBytecode: Bytes | undefined,
+        encoderOverrides: {
+            encodeInput: (fnName: string, values: any) => Bytes;
+            decodeOutput: (fnName: string, data: Bytes) => any;
+        },
+    ): T;
+    ABI(): ContractAbi | MethodAbi[];
+}
+
+export function getDeterministicContractAddressFromBytecode(bytecode: Bytes): Address {
+    return hexUtils.leftPad(hexUtils.hash(bytecode), ADDRESS_SIZE);
+}
+
+export function getBytecodeFromArtifact(artifact: ContractVersionData): Bytes {
+    return artifact.compilerOutput.evm.deployedBytecode.object;
+}
+
+export function getDeterministicContractAddressFromArtifact(artifact: ContractVersionData): Address {
+    return getDeterministicContractAddressFromBytecode(getBytecodeFromArtifact(artifact));
+}
+
+export interface GeneratedContract extends BaseContract {
+    getABIDecodedReturnData: <T>(methodName: string, returnData: string) => T;
+}
+
+export type ContractFunction<TArgs extends any[], TReturn> = (...args: TArgs) => ContractFunctionObj<TReturn>;
+
+export type ValueByChainId<T> = { [k in ChainId]: T };
+
+// TODO(kimpers): Consolidate this implementation with the one in @0x/token-metadata
+export function valueByChainId<T>(rest: Partial<ValueByChainId<T>>, defaultValue: T): { [key in ChainId]: T } {
+    // TODO I don't like this but iterating through enums is weird
+    return {
+        [ChainId.Mainnet]: defaultValue,
+        [ChainId.Ropsten]: defaultValue,
+        [ChainId.Rinkeby]: defaultValue,
+        [ChainId.Kovan]: defaultValue,
+        [ChainId.Ganache]: defaultValue,
+        [ChainId.BSC]: defaultValue,
+        [ChainId.Polygon]: defaultValue,
+        [ChainId.PolygonMumbai]: defaultValue,
+        ...(rest || {}),
+    };
+}
+
+export function createContractWrapperAndHelper<TContract extends GeneratedContract>(
+    chain: Chain,
+    contractType: ContractWrapperType<TContract>,
+    artifactName: string,
+    address?: Address,
+): [TContract, ContractHelper<TContract>] {
+    const fastAbi = new FastABI(
+        contractType.ABI() as MethodAbi[],
+        { BigNumber },
+    );
+    const artifact = (artifacts as ArtifactsMap)[artifactName];
+    const wrapper = new contractType(
+        address || getDeterministicContractAddressFromArtifact(artifact),
```

Comment:
Undeployed contracts will get a deterministic address based on its bytecode. This keeps the batcher from uploading multiple copies of the same contract, since they'll all collapse into a single override address.

Reply:
nice